File: 934.最短的桥.py
#
# @lc app=leetcode.cn id=934 lang=python3
#
# [934] 最短的桥
#

# @lc code=start
class Solution:
    def shortestBridge(self, grid: List[List[int]]) -> int:
        q = []
        m, n  = len(grid), len(grid[0])
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        
        def dfs(x, y):
            if not (0 <= x < m and 0 <= y < n and grid[x][y] == 1):
                return
            grid[x][y] = 0
            q.append((x, y))
            for direction in directions:
                dfs(x + direction[0], y + direction[1])
        
        visited = set()
        res = 0
        find = False
        for i in range(m):
            for j in range(n):
                if grid[i][j] and not find:
                    find = True
                    dfs(i, j)
        while q:
            for _ in range(len(q)):
                x, y = q[0]
                q = q[1:]
                for direction in directions:
                    curr_x, curr_y = x +direction[0], y + direction[1]

                    if not (0 <= curr_x < m and 0 <= curr_y < n) or (curr_x, curr_y) in visited:
                        continue
                    if grid[curr_x][curr_y]:
                        return res
                    else:
                        q.append((curr_x, curr_y))
                        visited.add((curr_x, curr_y))
            res += 1
        return res
# @lc code=end

File: 934.最短的桥.go
/*
 * @lc app=leetcode.cn id=934 lang=golang
 *
 * [934] 最短的桥
 */

// @lc code=start
func shortestBridge(grid [][]int) (res int) {
	q := [][2]int{}
	m, n := len(grid), len(grid[0])
	directions := [4][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}

	var dfs func(x, y int)
	dfs = func(x, y int) {
		if !(0 <= x && x < m && 0 <= y && y < n && grid[x][y] == 1) {
			return
		}
		grid[x][y] = 0
		q = append(q, [2]int{x, y})
		for _, direction := range directions {
			dfs(x+direction[0], y+direction[1])
		}
	}

	find := false
	for i := 0; i < m; i++ {
		if find {
			break
		}
		for j := 0; j < n; j++ {
			if grid[i][j] == 1 {
				find = true
				dfs(i, j)
				break
			}
		}
	}
	visited := map[[2]int]bool{}
	for len(q) > 0 {
		length := len(q)
		for i := 0; i < length; i++ {
			point := q[0]
			q = q[1:]
			for _, direction := range directions {
				x, y := point[0]+direction[0], point[1]+direction[1]
				if !(0 <= x && x < m && 0 <= y && y < n) || visited[[2]int{x, y}] {
					continue
				}
				if grid[x][y] == 1 {
					return
				} else {
					q = append(q, [2]int{x, y})
					visited[[2]int{x, y}] = true
				}
			}
		}
		res++
	}
	return
}

// @lc code=end

